sdk/client/client.ts

```diff
@@ -196,91 +248,83 @@ export class Client {
-
-    // Return a function
-    return function() {
-      // Collect the arguments passed into the wrapper to use later
-      var topLevelArguments = arguments
-      // Return a promise to ensure that the function remains 'thenable'
-      return new Promise(function(resolve, reject) {
+  private apiFunctionWrapper(apiFunction, api ) {
+
+    // Return a function, thus not immediately invoking
+
+    return ( ...args ) => {
+
+      // Return a promise to ensure that the function remains '.then()-able'
+      return new Promise( (resolve, reject) => {
+
         // Trigger a token refresh
-        var oauthPopulated = self.refreshToken(
-          self.authentications['oauth2'],
-          self.refreshLimit,
-          self.tokenUrlDetails,
-          self.usernameDetails,
-          self.passwordDetails,
-          self.clientIdDetails,
-          self.clientSecretDetails
-        )
-        // Once this has completed pass the oauth2 details on
-        oauthPopulated.then(function(oauth2Details: Oauth2) {
+        this.refreshToken(
+          this.authentications.oauth2,
+          this.refreshLimit,
+          this.tokenUrlDetails,
+          this.usernameDetails,
+          this.passwordDetails,
+          this.clientIdDetails,
+          this.clientSecretDetails
+        ).then( (oauth2Details: Oauth2) => {
+
           // Update the clients oauth2 details
-          self.authentications.oauth2 = oauth2Details
+          this.authentications.oauth2 = oauth2Details
           // Update the access token of the api being called
-          api['authentications']['oauth2']['accessToken'] = self.authentications.oauth2.accessToken
+          api.authentications.oauth2.accessToken = this.authentications.oauth2.accessToken
+
           /*
           Resolve the promise with the function that was wrapped
           In this case api is the api that this function is a part of,
           this is required to ensure that the function is called
           in the right context. The second argument topLevelArguments
           is the arguments passed into the Wrapper
           */
-          resolve(apiFunction.apply(api, topLevelArguments))
+
+          resolve( apiFunction.apply( api, args ) );
+
         })
         // Error handling
         .catch((err) => reject(err))
       })
     }
   }
 
-  /*
-  The fetchCredentials function gets the credentials from the specified source
-  */
-  private fetchCredentials(source: Source, value: string): string {
-
-    // Environment source
-    if (source == Source.Environment) {
-      // Set the credential using an environment variable
-      var credential: string = process.env[value] || 'MISSING'
-      // If it the environment variable does not exist then throw error
-      if (credential == 'MISSING') {
-        throw "Environment variable " + value + " has not been specified"
-      }
+  private fetchConfigurationItem( sourceType: Source, itemName: string ): string {
```

Comment:
Nice function! I like that it is not requiring the secrets file for each item as it was earlier.